Add API for creating `DocHandle`s using the unstable Automerge branch

acurrieclark opened the pull request:
This introduces a new method to the `DocCollection` class to allow new `DocHandle`s to be created using the Automerge `unstable` branch.

Tests were failing to build due to an incorrect type on the `DocHandle` change method and an import from `dist` in one of the tests. Corrected both of those here too, but please let me know if you would like me to separate those out into another PR.

pvh commented on packages/automerge-repo/src/DocCollection.ts:
```diff
@@ -29,6 +33,15 @@ export class DocCollection extends EventEmitter<DocCollectionEvents<unknown>> {
     return handle
   }
 
+  createUnstable<T>(): DocHandle<T> {
```
i'd rather not have a specific method for this -- i'm okay with passing in arguments somehow but because this Unstable API might not be there in other releases i don't think we should bake it in here.

pvh replied:
thinking more about this, i wonder if it would be better to expose a library configuration setting instead of doing this in automerge-repo, so instead of doing this you'd call something like Automerge.enableExperimental() or something... WDYT, @alexjg ?

acurrieclark replied:
Noted. Will adjust for so that an option is passed in to `create`.

I had initially considered passing it in as a global option when instantiating a `Repo`,  but wanted the option of having both `stable` and `unstable` documents. The same problem might occur if there was a separate library configuration option, as it would make it more difficult to chose on a per document level?

pvh commented:
Not to dispute it, but when would you want that?

On Thu, Feb 23, 2023 at 00:05 Alex Currie-Clark ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In packages/automerge-repo/src/DocCollection.ts
> <https://github.com/automerge/automerge-repo/pull/22#discussion_r1115341427>
> :
>
> > @@ -29,6 +33,15 @@ export class DocCollection extends EventEmitter<DocCollectionEvents<unknown>> {
>      return handle
>    }
>
> +  createUnstable<T>(): DocHandle<T> {
>
> Noted. Will adjust for so that an option is passed in to create.
>
> I had initially considered passing it in as a global option when
> instantiating a Repo, but wanted the option of having both stable and
> unstable documents. The same problem might occur if there was a separate
> library configuration option, as it would make it more difficult to chose
> on a per document level?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/automerge/automerge-repo/pull/22#discussion_r1115341427>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAWQD2AQ3KT6SIKNKTKFLWY4K6JANCNFSM6AAAAAAVERGM7U>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>
-- 
Peter van Hardenberg
"Everything was beautiful, and nothing hurt."—Kurt Vonnegut

acurrieclark commented:
To be fair, mostly for testing. I have a situation right now where I have an application which has been built on `stable` and I am in the process of migrating the document schema to `unstable`. Wanted to be able to instantiate both versions until I had made the transition.

I can see that there are perhaps few real world requirements outside of dev/testing.

acurrieclark commented:
Just realised that if this is to work, I need to add a test to ensure that unstable documents can be found too. Doing that now

acurrieclark commented:
Unstable flag now added to the `find` method too, and a test added.

While implementing the test, I noticed that the `Repo` test `can find a created document` is probably not testing what we want here. As it is looking for a `handle` which already exists in the repo, it simply loads the `handle` from the `DocCollection` cache, without touching the network or storage adapters. The test I added for `unstable` makes a fresh repo and reuses the storage adapter.

acurrieclark commented:
Been thinking about how I personally will use this, and realised that I can make it work either way; passing an option on a per document basis or globally to the Repo on instantiation. The global option is certainly more straightforward, and I can see how it could also happen at the `automerge` level.

Either way, I would love to be able to use `automerge-repo` with the `unstable` branch, so let me know if you have a preference and I will happily put together the PR.

acurrieclark commented:
Closing this for now. Will rethink its implementation further down the line.
